CIP-????/README.md

```diff
@@ -0,0 +1,676 @@
+
+### Costs
+
+The costs of this proposal are substantial.
+
+Plutus Core today has only 8 kinds of term, this proposal adds 2 more, an increase of 25%.
+Additionally, the new term types are the first term types which have a variable number of children.
+Both of these changes increase implementation complexity throughout the system: everything that processes Plutus Core terms must now handle the new cases, and handle the variable numbers of children.
+
+This change also modestly expends our novelty budget.
+Plutus Core is a deliberately conservative language: for the most part it is just the untyped lambda calculus.
+The proposed new features for sums-of-products are also conservative, and follow a very typical pattern.
+But they are not _quite_ as standard as the untyped lambda calculus.
+
+Moreover, a change like this would be very painful to reverse, so we should seriously consider the costs before proceeding.
+
+### Discussion
+
+#### Why is case not strict?
+
+This makes it different to all other Plutus Core terms, but we believe it is the expected behaviour: _no_ language has strict case destructuring.
+
+Consider e.g.
+
+```haskell
+case (Just 1) of 
+   Just x -> print "Success!"
+   Nothing -> error "Failure!"
+```
+
+Even in a strict language like Rust, nobody would expect this to evaluate the `Nothing` branch and evaluate to an error.
+
+Furthermore, this additional laziness has significant performance benefits, see Appendix 2 for more discussion.
+
+#### Is this still worth it if users aren't decoding the script context?
+
+The issues with decoding the script context only apply to users who actually try and decode it; many people have moved on to just using the underlying Data object instead.
+Arguably that saps the motivation for this proposal.
+
+There are two responses to this:
+1. This proposal seems worth it on the basis of its other benefits (benefits 2 and 3).
+2. Removing the problems with decoding the script context may allow users to return to that approach, which can be more ergonomic (and potentially cheaper, since builtin calls are expensive).
+
+#### Why not also make datums and redeemers terms instead of `Data`?
+
+It is always nice to get rid of things, and this CIP does _not_ let us get rid of `Data`, because it is still used for redeemers and datums.
+That begs the question: could we also have redeemers and datums be Plutus Core terms instead of values of type `Data`?
+
+The problem is that, unlike the encoded script context, redeemers and datums are _user facing_.
+They appear in the block format, they can be supplied by users (e.g. directly at the CLI!), and they can be inspected in block explorers.
+There is therefore a significant benefit to having them in a simpler format that is easier to work with.
+If they were Plutus Core terms in generality, we could have all kinds of strange things like people embedding functions in them.
```

Comment:
I've thought about this for a while, but maybe this would be a good thing?
You could maintain a transformation from Data to SoP as I noted elsewhere to keep backwards compatibility (and should remove Data from UPLC in any case IMO),
but allowing arbitrary terms, i.e. essentially functions, in datums would open up many use cases.

For example, you could split a large script into many smaller UTXOs, and keep them in the _datums_ as functions.
Then a small script remains that reads the functions from datums and evaluates them.

Other use cases might be protocols that allow you to add arbitrary extra logic through passing in functions.

What do you think?

Reply:
This all comes at the cost of destroying intepretability. I do think that's valuable.

I think a good comparison is with serialization. Typically you can't serialize functions. This is sometimes annoying, but you can usually get around it by defining a data type that represents the functions you care about that _is_ serializable (effectively: writing your own little language). This is not as powerful as being able to serialize functions, but it does the job and doesn't sacrifice interpretability.

Now, there might be other potential benefits e.g. sharing of script fragments, but I don't think we should sacrifice _everything_ on the altar of performance, and I do think that interpretability is important for redeemers and datums.

Reply:
The same way you encode any term/script; it's no different.